### PR TITLE
Optical flow: add RAFT detector

### PR DESCRIPTION
**Pull request by danielmend**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -109,7 +109,8 @@ This module exposes a single function `download` which takes the same arguments
 * **cut_framerates** a list of additional framerates to detect jump cuts at. If None, jump cuts will only be detected at the original framerate of the video (default *None*)
 * **cuts_are_clips** whether or not to turn each contiguous section of each input video into a distinct ouput video (default *False*)
 * **stage** which stage of processing to execute in betweeen downloading + cheap subsampling and costly subsampling (default *"download"*)
-* **optical_flow_detector** Which optical flow detector to use (default *"cv2"*)
+* **optical_flow_detector** Which optical flow detector (can be "cv2" or "raft") to use (default *"cv2"*)
+* **optical_flow_params** Tuple of hyperparameters to modify optical flow detector (default *None*)
```
Comment (iejMac) [inline README.md:113]: this is confusing, too many params about optical flow also optical_flow_params is too vague. Let's just choose one direction and go with that. I suggest one dict param: optical_flow_params and putting everything in there - detector, fps, downsample size etc.

also come up with some reasonable defaults for this

File: video2dataset/subsamplers/optical_flow_subsampler.py [hunk 2]
```diff
@@ -4,6 +4,92 @@
...
+    # Resize the image using the calculated dimensions
+    resized_image = cv2.resize(image, (new_width, new_height), interpolation=cv2.INTER_AREA)
+
+    return resized_image, scaling_factor
+
+class RAFTDetector:
+    def __init__(
+        self, args, downsample_size=None, device='cuda'
+    ):
+        model = torch.nn.DataParallel(RAFT(args))
```
Comment (iejMac) [inline video2dataset/subsamplers/optical_flow_subsampler.py:59]: why are you using DataParallel here?

File: video2dataset/subsamplers/optical_flow_subsampler.py [hunk 3]
```diff
@@ -4,6 +4,92 @@
...
+
+        model = model.module
+        model.to(device)
+        model.eval()
+
+        self.model = model
+        self.device = device
+        self.downsample_size = downsample_size
+
+    def preprocess(self, frame1, frame2):
```
Comment (iejMac) [inline video2dataset/subsamplers/optical_flow_subsampler.py:70]: you already said results are good I think but always double triple check if prepro is correct

File: video2dataset/subsamplers/optical_flow_subsampler.py (hunk 2 above)
Comment (danielmend) [reply video2dataset/subsamplers/optical_flow_subsampler.py:63]: the raft model wont load without it lol

File: video2dataset/subsamplers/optical_flow_subsampler.py [hunk 4]
```diff
@@ -120,15 +217,20 @@ class OpticalFlowSubsampler:
...
+                pyr_scale, levels, winsize, iterations, poly_n, poly_sigma, flags = args
                 self.detector = Cv2Detector(
                     pyr_scale, levels, winsize, iterations, poly_n, poly_sigma, flags, downsample_size=downsample_size
                 )
             else:
                 self.detector = Cv2Detector(downsample_size=downsample_size)
+        elif detector == "raft":
+            assert args is not None
+            if not isinstance(args, AttrDict):
+                args = AttrDict(args)
```
Comment (danielmend) [inline video2dataset/subsamplers/optical_flow_subsampler.py:232]: this is required to make RAFT work (it expects command line args...) i can either refactor RAFT on my fork or include this quick fix (i prefer the latter)

Comment (iejMac) [reply video2dataset/subsamplers/optical_flow_subsampler.py:233]: nope this is good

File: video2dataset/subsamplers/optical_flow_subsampler.py (hunk 2 above)
Comment (iejMac) [reply video2dataset/subsamplers/optical_flow_subsampler.py:89]: what do you mean? the RAFT model needs to be wrapped in a DataParallel?

Comment (iejMac) [reply video2dataset/subsamplers/optical_flow_subsampler.py:89]: are you saying that's how they saved the state dict?

**Comment by iejMac** [comment]
if this works end2end and lint tests pass I'm fine with merging @danielmend

File: video2dataset/main.py [hunk 5]
```diff
@@ -62,6 +62,7 @@ def video2dataset(
     cuts_are_clips: bool = False,
     encode_formats: dict = None,
     stage: str = "download",
+    optical_flow_params: dict = None,
     optical_flow_detector: str = "cv2",
```
Comment (iejMac) [inline video2dataset/main.py:66]: well we don't want these params right?

File: video2dataset/slurm_executor.py [hunk 6]
```diff
@@ -20,10 +20,11 @@ def __init__(self, global_task_id, num_tasks):
         self.num_tasks = num_tasks
 
     def __call__(self, shardfile_list):
+        print(shardfile_list, flush=True)
```
Comment (iejMac) [inline video2dataset/slurm_executor.py:23]: should this be here?

**Pull request merged**